doc/developer/design/20230531_system_privileges.md

```diff
@@ -0,0 +1,97 @@
+`GRANT SELECT ON TABLE t TO r` will grant the ability to read table `t` to role `r`. `GRANT CREATE
+ON SCHEMA s TO r` will grant the ability to create new objects in schema `s` to role `r`. Object
+privileges can be inherited through role membership. Role attributes control the user's ability to
+create (and sometimes modify) top level objects, they are granted and revoked with the `ALTER ROLE`
+keywords. For example, `ALTER ROLE r CREATEDB` will grant the ability to create databases to
+role `r`. `ALTER ROLE r CREATECLUSTER` will grant the ability to create clusters to role `r`. Role
+attributes cannot be inherited through role membership.
+
+Role attributes always exist in a boolean state. To change the state of a role attribute you must
+alter the role. For example `ALTER ROLE joe CREATEDB` sets the `CREATEDB` attribute to `true`,
+and `ALTER ROLE NOCREATEDB` sets the `CREATEDB` attribute to `false`.
+
+On the other hand, privileges either exist or they don't exist. You create them by granting a
+privilege via `GRANT ...` and you destroy them via `REVOKE ...`.
+
+Role attributes present some problems as they currently exist:
+
+- It can be confusing that there are two separate privilege primitives that control a role's
+  privilege, and they have different semantics.
+- Role attributes are not inherited which makes it cumbersome to grant attributes to a large
+  number of roles.
+- Users are not able to easily grant the ability to create top level objects to many current and
+  future users in a short amount of time.
+
+## Goals
+
+- Reduce confusion around privileges.
+- Allow the ability to create databases, clusters, and roles to be inherited through role
+  membership.
+- Allow users to easily grant the ability to create top level objects to many current and future
+  users in a short amount of time.
+- Maintain PostgreSQL compatibility.
+
+## Overview
+
+We will add a new type of privilege called System Privileges. They will look and act just like
+normal object privileges, but they will grant the abilities that would otherwise be granted through
+role attributes. They will exist in parallel with role attributes.
+
+NOTE: Other databases, such as CockroachDB and Snowflake, have some form of system privileges that
+control similar privileges to role attributes.
+
+## Detailed description
+
+System privileges will be granted with the following syntax:
+`GRANT <privilege-specification> ON SYSTEM TO <role-specification>`
+
+System privileges will be revoked with the following syntax:
+`REVOKE <privilege-specification> ON SYSTEM FROM <role-specification>`
+
+System privileges and role attributes will interact in the following ways:
+
+- Granting a system privilege will grant the privilege AND set the attribute to true.
+- Revoking a system privilege will revoke the privilege AND set the attribute to false.
+- Altering a role attribute will affect the role attribute but will not change the system privilege.
```

Comment:
> Altering a role attribute will affect the role attribute but will not change the system privilege.

I'm assuming we want this for postgres compatibility? For the cases when a user changes a role attribute in a way that disagrees with the system privilege, can we emit a notice? e.g. if someone runs `ALTER ROLE parker NOCREATEDB`, but I still have the ability to create a database via my system privileges, can we emit a warning indicating that?

Reply:
> I'm assuming we want this for postgres compatibility?

Yes, if someone execute `ALTER ROLE r CREATEDB` and `r` is granted the system privilege `CREATEDB`, then all members of `r` will also implicitly be granted the system privilege `CREATEDB`. This effectively makes role attributes inheritable, which would not be compatible with PostgreSQL. I'll add this detail to the doc.

I think the notice is a good idea, I'll add it to the doc too.

Reply:
Great!